File: web/src/import/collection-review.ts
import type {
  LibraryReview,
  PlaylistReview,
  SpotifyCollectionReview,
} from '../api/client'
import type { ListeningExportSnapshot } from './snapshot'
export type FileSelection = {
  ordinal: number
  name: string
  role: 'playlist' | 'liked' | 'skip'
  target: string | null
  createKey: string
  fileHash: string
}
export type CollectionSelection = {
  context: SpotifyCollectionReview
  files: FileSelection[]
  libraryMode: 'add' | 'replace'
  confirmed: boolean
  confirmRemovals: boolean
}
export function initialSelection(
  snapshot: ListeningExportSnapshot,
  context: SpotifyCollectionReview,
): CollectionSelection {
  return {
    context,
    confirmed: false,
    confirmRemovals: false,
    libraryMode: 'add',
    files: snapshot.playlists.map((p, i) => {
      const matches = context.playlists.filter((t) => t.fileHash === p.key)
      return {
        ordinal: p.ordinal,
        name: p.name,
        role: 'playlist',
        target: matches.length === 1 ? matches[0].key : null,
        createKey: `exportify:${p.key}:${i}`,
        fileHash: p.key,
      }
    }),
  }
}
export function applySelection(
  original: ListeningExportSnapshot,
  selection: CollectionSelection,
): {
  snapshot: ListeningExportSnapshot
  libraryReview: LibraryReview
  playlistReview: PlaylistReview
} {
  if (
    !selection.confirmed ||
    selection.files.filter((f) => f.role === 'liked').length > 1
  )
    throw new Error('Review the collection roles before importing.')
  const library = new Map(original.library.map((r) => [r.platformId, r]))
  const playlists: ListeningExportSnapshot['playlists'] = [],
    playlistReview: PlaylistReview = []
  const used = new Set<string>()
  for (const file of selection.files) {
    if (file.role === 'skip') continue
    const p = original.playlists.find((p) => p.ordinal === file.ordinal)
    if (!p) throw new Error('Review the files again.')
    if (file.role === 'liked') {
      for (const e of p.entries)
        if (e.platformId)
          library.set(e.platformId, {
            platformId: e.platformId,
            playCount: null,
            skipCount: null,
            lastPlayedAt: null,
            dateAdded:
              e.addedAt ?? library.get(e.platformId)?.dateAdded ?? null,
            likeRating: null,
          })
    } else {
      const target =
        file.target === null
          ? null
          : selection.context.playlists.find((t) => t.key === file.target)
      if (file.target !== null && !target)
        throw new Error('Choose a current playlist.')
      const key = target?.key ?? file.createKey
      if (used.has(key) || !file.name.trim() || file.name.length > 500)
        throw new Error('Choose distinct playlists and valid names.')
      used.add(key)
      playlists.push({
        ...p,
        ordinal: playlists.length,
        key,
        name: file.name.trim(),
      })
      playlistReview.push({
        key,
        baseFingerprint: target?.fingerprint ?? null,
        fileHash: file.fileHash,
      })
    }
  }
  const hasLikes =
    original.package === 'spotify_account' ||
    selection.files.some((f) => f.role === 'liked')
  if (
    (!hasLikes && playlists.length === 0) ||
    (original.package === 'spotify_exportify' &&
      library.size === 0 &&
      playlists.every((p) => p.entries.length === 0))
  )
    throw new Error('Choose at least one collection.')
  if (
    selection.libraryMode === 'replace' &&
    hasLikes &&
    !selection.confirmRemovals
  )
    throw new Error('Confirm replacing imported Liked Songs.')
  const libraryReview: LibraryReview =
    selection.libraryMode === 'replace' && hasLikes
      ? { mode: 'replace', fingerprint: selection.context.library.fingerprint }
      : { mode: 'add' }
  const ids = new Set([
    ...library.keys(),
    ...playlists.flatMap((p) =>
      p.entries.flatMap((e) => (e.platformId ? [e.platformId] : [])),
    ),
  ])
  return {
    snapshot: {
      ...original,
      tracks: original.tracks.filter((t) => ids.has(t.platformId)),
      library: [...library.values()],
      unresolved:
        original.package === 'spotify_exportify'
          ? {
              rows: original.playlists
                .filter((p) =>
                  selection.files.some(
                    (f) => f.ordinal === p.ordinal && f.role !== 'skip',
                  ),
                )
                .reduce(
                  (n, p) => n + p.entries.filter((e) => !e.platformId).length,
                  0,
                ),
              plays: 0,
            }
          : original.unresolved,
      playlists,
    },
    libraryReview,
    playlistReview,
  }
}

File: client/lib/import/collection_review.dart
import '../data/listening/listening_models.dart';
import 'snapshot.dart';

class ImportedCollection {
  const ImportedCollection({
    required this.key,
    required this.name,
    required this.fingerprint,
    required this.fileHash,
  });
  final String key, name, fingerprint;
  final String? fileHash;
}

class CollectionContext {
  const CollectionContext({
    required this.ids,
    required this.fingerprint,
    required this.playlists,
    this.hasQuickImport = false,
  });
  final bool hasQuickImport;
  final List<String> ids;
  final String fingerprint;
  final List<ImportedCollection> playlists;
  factory CollectionContext.fromJson(Map<String, dynamic> json) {
    final library = asObject(json['library']);
    return CollectionContext(
      hasQuickImport: json['hasQuickImport'] == true,
      ids: asList(library['ids']).cast<String>(),
      fingerprint: readString(library, 'fingerprint'),
      playlists: asList(json['playlists']).map((value) {
        final p = asObject(value);
        return ImportedCollection(
          key: readString(p, 'key'),
          name: readString(p, 'name'),
          fingerprint: readString(p, 'fingerprint'),
          fileHash: p['fileHash'] as String?,
        );
      }).toList(),
    );
  }
}

class CollectionFile {
  const CollectionFile({
    required this.ordinal,
    required this.name,
    required this.role,
    required this.target,
    required this.createKey,
    required this.fileHash,
  });
  final int ordinal;
  final String name, role, createKey, fileHash;
  final String? target;
  CollectionFile change({
    String? name,
    String? role,
    String? action,
    String? newKey,
  }) => CollectionFile(
    ordinal: ordinal,
    name: name ?? this.name,
    role: role ?? this.role,
    target: action == null
        ? target
        : action.isEmpty
        ? null
        : action,
    createKey: newKey ?? createKey,
    fileHash: fileHash,
  );
}

class CollectionSelection {
  const CollectionSelection({
    required this.context,
    required this.files,
    this.mode = 'add',
    this.confirmed = false,
    this.confirmRemovals = false,
  });
  final CollectionContext context;
  final List<CollectionFile> files;
  final String mode;
  final bool confirmed, confirmRemovals;
  factory CollectionSelection.initial(
    ListeningExportSnapshot snapshot,
    CollectionContext context,
  ) => CollectionSelection(
    context: context,
    files: snapshot.playlists.map((p) {
      final matches = context.playlists
          .where((t) => t.fileHash == p.key)
          .toList();
      return CollectionFile(
        ordinal: p.ordinal,
        name: p.name,
        role: 'playlist',
        target: matches.length == 1 ? matches.single.key : null,
        createKey: 'exportify:${p.key}:${p.ordinal}',
        fileHash: p.key,
      );
    }).toList(),
  );
  CollectionSelection change({
    List<CollectionFile>? files,
    String? mode,
    bool? confirmed,
    bool? confirmRemovals,
  }) => CollectionSelection(
    context: context,
    files: files ?? this.files,
    mode: mode ?? this.mode,
    confirmed: confirmed ?? this.confirmed,
    confirmRemovals: confirmRemovals ?? this.confirmRemovals,
  );
  bool canUpload(ListeningExportSnapshot snapshot) {
    try {
      apply(snapshot);
      return true;
    } catch (_) {
      return false;
    }
  }

  ({
    ListeningExportSnapshot snapshot,
    Map<String, Object?> libraryReview,
    List<Map<String, Object?>> playlistReview,
  })
  apply(ListeningExportSnapshot original) {
    if (!confirmed || files.where((f) => f.role == 'liked').length > 1) {
      throw const FormatException('Review collection roles.');
    }
    final library = {for (final row in original.library) row.platformId: row};
    final playlists = <SnapshotPlaylist>[],
        review = <Map<String, Object?>>[],
        keys = <String>{};
    for (final file in files) {
      if (file.role == 'skip') continue;
      final p = original.playlists.firstWhere((p) => p.ordinal == file.ordinal);
      if (file.role == 'liked') {
        for (final e in p.entries) {
          if (e.platformId != null) {
            library[e.platformId!] = SnapshotLibraryRow(
              platformId: e.platformId!,
              dateAdded: e.addedAt ?? library[e.platformId!]?.dateAdded,
            );
          }
        }
      } else {
        final target = file.target == null
            ? null
            : context.playlists.firstWhere((t) => t.key == file.target);
        final key = target?.key ?? file.createKey;
        if (!keys.add(key) ||
            file.name.trim().isEmpty ||
            file.name.length > 500) {
          throw const FormatException(
            'Choose distinct playlists and valid names.',
          );
        }
        playlists.add(
          SnapshotPlaylist(
            ordinal: playlists.length,
            key: key,
            name: file.name.trim(),
            description: p.description,
            lastModifiedAt: p.lastModifiedAt,
            entries: p.entries,
          ),
        );
        review.add({
          'key': key,
          'baseFingerprint': target?.fingerprint,
          'fileHash': file.fileHash,
        });
      }
    }
    final hasLikes =
        original.package == ExportPackage.spotifyAccount ||
        files.any((f) => f.role == 'liked');
    if ((!hasLikes && playlists.isEmpty) ||
        (original.package == ExportPackage.spotifyExportify &&
            library.isEmpty &&
            playlists.every((p) => p.entries.isEmpty))) {
      throw const FormatException('Choose a collection.');
    }
    if (mode == 'replace' && hasLikes && !confirmRemovals) {
      throw const FormatException('Confirm replacing Liked Songs.');
    }
    final ids = {
      ...library.keys,
      ...playlists
          .expand((p) => p.entries)
          .map((e) => e.platformId)
          .whereType<String>(),
    };
    return (
      snapshot: ListeningExportSnapshot(
        package: original.package,
        timeZone: original.timeZone,
        country: original.country,
        tracks: original.tracks
            .where((t) => ids.contains(t.platformId))
            .toList(),
        days: original.days,
        library: library.values.toList(),
        artists: original.artists,
        playlists: playlists,
        unresolved: original.package == ExportPackage.spotifyExportify
            ? SnapshotUnresolved(
                rows: original.playlists
                    .where(
                      (p) => files.any(
                        (f) => f.ordinal == p.ordinal && f.role != 'skip',
                      ),
                    )
                    .expand((p) => p.entries)
                    .where((e) => e.platformId == null)
                    .length,
                plays: 0,
              )
            : original.unresolved,
        ledgerFrom: original.ledgerFrom,
        ledgerTo: original.ledgerTo,
      ),
      libraryReview: mode == 'replace' && hasLikes
          ? {'mode': 'replace', 'fingerprint': context.fingerprint}
          : {'mode': 'add'},
      playlistReview: review,
    );
  }
}
